internal/search/query/parser.go

```
@@ -183,6 +183,57 @@ func (p *parser) skipSpaces() error {
 	return nil
 }
 
+// ScanDelimited takes a delimited (e.g., quoted) value for some arbitrary
+// delimiter, returning the undelimited value, and the end position of the
+// original delimited value (i.e., including quotes). `\` is treated as an
+// escape character for the delimiter and traditional string escape sequences.
+// The input buffer must start with the chosen delimiter.
+func ScanDelimited(buf []byte, delimiter rune) (string, int, error) {
+	var count, advance int
+	var r rune
+	result := []rune{}
+
+	next := func() {
+		r, advance = utf8.DecodeRune(buf)
+		count += advance
+		buf = buf[advance:]
+	}
+
+	next()
+	if r != delimiter {
+		panic(fmt.Sprintf("ScanDelimited expects the input buffer to start with delimiter %s, but it starts with %s.", string(delimiter), string(r)))
```

What makes you panic here rather than returning an error?

---

Not sure if this is useful, but definitely related: https://golang.org/pkg/strconv/#UnquoteChar

---

It's in fact what https://golang.org/pkg/strconv/#Unquote uses under the hood.

---

Oh, you mean `UnquoteChar` for printing the error?

As for panic: the intent is to enforce a function specification, that it is only to be used in a certain way _internal_ to our implementation (i.e., if someone wanted to call this function for other purposes). Similar to how we do [for pagination](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@ee07341a8b1749f5bc7d294cf82889554dd1650a/-/blob/cmd/frontend/graphqlbackend/search_pagination.go#L93) and probably other places. I want the returned `error` to retain the meaning that something went wrong during scanning/parsing, which could possibly be used to surface an application-level error.

There's the possibility of using different error return types, and then checking these in some caller, but there's no guarantee others others will do the same thing (they might just propagate the error up without checking). It may turn out that this function doesn't need to be exposed outside of the package, in which case the panic is overkill. I'll probably revisit this need at a later stage.